src/client_side_reply.cc

```diff
@@ -1486,15 +1487,14 @@ clientReplyContext::identifyStoreObject()
 
     // client sent CC:no-cache or some other condition has been
     // encountered which prevents delivering a public/cached object.
-    // XXX: The above text does not match the condition below. It might describe
-    // the opposite condition, but the condition itself should be adjusted
-    // (e.g., to honor flags.noCache in cache manager requests).
-    if (!r->flags.noCache || r->flags.internal) {
+    if (r->flags.noCache) {
```

Comment:
> PR description:  This case was spotted (with an XXX) but not fixed in that revision.

This XXX is much broader than the former PR title and the current PR description imply. It is not really about cache manager requests. Cache manager requests is just a simple _example_ of what is wrong with this code. It could be the only example in the current code, but that does not really matter for that XXX true scope. We will adjust the PR description when the dust settles.

For now, I have one primary concern: Is the following combination of conditions possible? If yes, what happens to requests in those cases?

* flags.internal is true
* flags.noCache is true
* the request is _not_ a cache manager request

I have not checked whether the following is an example of the above, but I am thinking of an HTTP request for this Squid MIME icon that has a Cache-Control: no-cache or a similar HTTP header. However, please do not let this example limit the scope of your investigation when answering the above question.

Reply:
> Is the following combination of conditions possible?

Yes, it is possible - I was able to reproduce this scenario requesting a MIME icon, with a request containing Cache-Control: no-cache. PR Squid returns 404 in this case (after calling internalStart() and proceeding to its last 'else'). This leads to removing of the cached MIME icon entry from the cache (before proceeding to the MISS path), and the subsequent requests to the same icon (even without no-cache) also result in 404.

I have not tested, but it looks like that a similar scenario should be possible for store_digest requests as well ("/squid-internal-periodic/store_digest"). In this case, internalStart() would treat this situation as "This cache is currently building its digest" state and return 404. However, the entry is going be replaced again after a genuine digest rebuild so that subsequent requests (without no-cache) would not result in 404.

Reply:
I tried to adjust this condition, so that it

Reply:
Thank you for this research and adjustments. I am resolving this change request, but we will use the information you have provided to finalize this PR changes.

Reply:
> hat a similar scenario should be possible for store_digest 

This situation is also possible for netdb entries (these entries are cached as common, "external" entries). I ran the similar test with "no-store" for store_digest and netdb entries.